vpr/src/place/analytical_placement/AnalyticalSolver.cpp

```diff
@@ -162,21 +211,255 @@ void QPHybridSolver::solve(unsigned iteration, PartialPlacement &p_placement) {
+                    tripletList_x.emplace_back(first_node_id, first_node_id, w);
+                    b_x(first_node_id) += w * p_placement.node_loc_x[second_node_id];
+                }else{
+                    tripletList_y.emplace_back(first_node_id, first_node_id, w);
+                    b_y(first_node_id) += w * p_placement.node_loc_y[second_node_id];
+                }
+            }
+        };
+        for (size_t node_id = 0; node_id < node_ids_no_duplicate.size(); node_id++) {
+            if (node_id != maxXId && node_id != minXId) {
+                add_node(node_id, maxXId, node_ids_no_duplicate.size(), true);
+                add_node(node_id, minXId, node_ids_no_duplicate.size(), true);
+            } 
+            if (node_id != maxYId && node_id != minYId) {
+                add_node(node_id, maxYId, node_ids_no_duplicate.size(), false);
+                add_node(node_id, minYId, node_ids_no_duplicate.size(), false);
+            } 
+        }
+        add_node(maxXId, minXId, node_ids_no_duplicate.size(), true);
+        add_node(maxYId, minYId, node_ids_no_duplicate.size(), false);
+    }
+    A_sparse_x.setFromTriplets(tripletList_x.begin(), tripletList_x.end());
+    A_sparse_y.setFromTriplets(tripletList_y.begin(), tripletList_y.end());
+}
+
+
+void B2BSolver::populate_matrix_anchor(PartialPlacement& p_placement, unsigned iteration) {
+    double epsilon = 1e-6;
+    double coeff_pseudo_anchor = 0.001 * std::exp((double)iteration/29.0);
+    // double coeff_pseudo_anchor = std::exp((double)iteration/1.0);
+    for (size_t i = 0; i < p_placement.num_moveable_nodes; i++){
+        // Anchor node are always 2 pins.
+        double pseudo_w_x = coeff_pseudo_anchor*2.0/std::max(std::abs(p_placement.node_loc_x[i] - node_loc_x_legalized[i]), epsilon);
+        double pseudo_w_y = coeff_pseudo_anchor*2.0/std::max(std::abs(p_placement.node_loc_y[i] - node_loc_y_legalized[i]), epsilon);
+        A_sparse_x.coeffRef(i, i) += pseudo_w_x;
+        A_sparse_y.coeffRef(i, i) += pseudo_w_y;
+        b_x(i) += pseudo_w_x * node_loc_x_legalized[i];
+        b_y(i) += pseudo_w_y * node_loc_y_legalized[i];
+    }
+    // Clique formulation anchor points
+    // double coeff_pseudo_anchor = 0.01 * std::exp((double)iteration/5.0);
+    // for (size_t i = 0; i < p_placement.num_moveable_nodes; i++){
+    //     double pseudo_w = coeff_pseudo_anchor; 
+    //     A_sparse_x.coeffRef(i, i) += pseudo_w;
+    //     A_sparse_y.coeffRef(i, i) += pseudo_w;
+    //     b_x(i) += pseudo_w * node_loc_x_legalized[i];
+    //     b_y(i) += pseudo_w *  node_loc_x_legalized[i];
+    // }
+}
+
+// This function is the inner loop that alternate between building matrix based on placement, and solving model to get placement.
+void B2BSolver::b2b_solve_loop(unsigned iteration, PartialPlacement &p_placement){
+    double previous_hpwl, current_hpwl = std::numeric_limits<double>::max();
+    unsigned counter = 0;
+    do{
+        counter++; 
```

Comment:
More of a nitpick, I would increment the counter at the end of the while loop. But this while loop is really just being used as a for loop. I would use a for loop over a while loop to make it more clear.